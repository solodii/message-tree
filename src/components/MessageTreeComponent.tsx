import * as React from 'react';
import { MessageTree } from '../state/types';
import { ROOT_SUBJECT } from '../state/constants';
import ReplyFormContainer from '../containers/ReplyFormContainer';
import MessageComponent from './MessageComponent';

type ReplyCallback = (e: React.MouseEvent<HTMLAnchorElement>) => void;

export interface Props {
  messageTrees: MessageTree[];
  onReply: (subjectId: string) => void;
}

export default class MessageTreeComponent extends React.Component<Props, {}> {
  public render(): JSX.Element {
    const treeCmp = this.renderTrees(
      this.props.messageTrees
    );
    return (
      <div>
        <ReplyFormContainer subjectId={ROOT_SUBJECT} />
        {treeCmp}
      </div>
    );
  }

  private renderTrees(messageTrees: MessageTree[]): JSX.Element {
    const trees = messageTrees.map((tree, index) => {
      const {
        message,
        hasDraft,
        children
      } = tree;
      const replyFormOrLink = hasDraft
        ? this.renderReplyForm(message.id)
        : this.renderReplyLink(message.id);
      const subtrees = this.renderTrees(children);
      return (
        <div key={index}>
          <MessageComponent message={message} />
          {replyFormOrLink}
          {subtrees}
        </div>
      );
    });
    return <div className="message-trees">{trees}</div>;
  }

  private renderReplyForm(subjectId: string): JSX.Element {
    return <ReplyFormContainer subjectId={subjectId} />;
  }

  private renderReplyLink(subjectId: string): JSX.Element {
    return (
      <a
        className="reply"
        href="#"
        onClick={this.onReplyClick(subjectId)}
      >
        reply
      </a>
    );
  }

  private onReplyClick(subjectId: string): ReplyCallback {
    return (e: React.MouseEvent<HTMLAnchorElement>) => {
      e.preventDefault();
      this.props.onReply(subjectId);
    };
  }
}
