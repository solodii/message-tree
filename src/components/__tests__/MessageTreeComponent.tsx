import * as React from 'react';
import { shallow } from 'enzyme';
import {
  messagesWithoutDraft,
  messageTrees
} from '../../mocks/data';
import MessageTreeComponent, { Props } from '../MessageTreeComponent';

describe('MessageTreeComponent without messageTrees', () => {
  const onReply = jest.fn();
  const initialProps: Props = {
    messageTrees: [],
    onReply
  };
  const wrapper = shallow<Props, {}>(
    <MessageTreeComponent {...initialProps} />
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('onReply is not called', () => {
    expect(onReply).toHaveBeenCalledTimes(0);
  });
});

describe('MessageTreeComponent with messageTrees', () => {
  const onReply = jest.fn();
  const initialProps: Props = {
    messageTrees,
    onReply
  };
  const wrapper = shallow<Props, {}>(
    <MessageTreeComponent {...initialProps} />
  );

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('onReply is called for each message without reply draft', () => {
    wrapper.find('.reply')
      .forEach((node) =>
        node.simulate('click', { preventDefault: jest.fn() })
      );
    const { calls } = onReply.mock;
    const accordance = messagesWithoutDraft
      .every((msg) =>
        calls.filter((args) => args[0] === msg.id).length === 1
      );
    expect(accordance).toBeTruthy();
  });
});
