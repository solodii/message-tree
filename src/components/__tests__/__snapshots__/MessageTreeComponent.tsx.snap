// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageTreeComponent with messageTrees renders correctly 1`] = `
<div>
  <ReplyFormContainer
    subjectId=""
  />
  <div
    className="message-trees"
  >
    <div>
      <MessageComponent
        message={
          Object {
            "date": 2016-11-10T08:01:00.000Z,
            "id": "0a6a7bfd-a212-4ae1-8125-6c7486213a94",
            "subjectId": "",
            "text": "first",
          }
        }
      />
      <ReplyFormContainer
        subjectId="0a6a7bfd-a212-4ae1-8125-6c7486213a94"
      />
      <div
        className="message-trees"
      >
        <div>
          <MessageComponent
            message={
              Object {
                "date": 2016-11-10T08:03:00.000Z,
                "id": "913d97b3-6e04-4c1e-a11c-222221abf00e",
                "subjectId": "0a6a7bfd-a212-4ae1-8125-6c7486213a94",
                "text": "third",
              }
            }
          />
          <a
            className="reply"
            href="#"
            onClick={[Function]}
          >
            reply
          </a>
          <div
            className="message-trees"
          />
        </div>
      </div>
    </div>
    <div>
      <MessageComponent
        message={
          Object {
            "date": 2016-11-10T08:02:00.000Z,
            "id": "6c6f1976-ff5b-420b-bc9d-f7484f0ee53f",
            "subjectId": "",
            "text": "second",
          }
        }
      />
      <a
        className="reply"
        href="#"
        onClick={[Function]}
      >
        reply
      </a>
      <div
        className="message-trees"
      />
    </div>
  </div>
</div>
`;

exports[`MessageTreeComponent without messageTrees renders correctly 1`] = `
<div>
  <ReplyFormContainer
    subjectId=""
  />
  <div
    className="message-trees"
  />
</div>
`;
